File: Data Structures & Algorithms/validate-parentheses/submission-20.py
class Solution:
    def isValid(self, s: str) -> bool:
        while '[]' in s or '{}' in s or '()' in s:
            s = self.myReplace(s, '[]', '')
            s = self.myReplace(s, '{}', '')
            s = self.myReplace(s, '()', '')
        
        return s == ''

    def myReplace(self, s: str, old: str, new: str):
        i = 0
        res = ""

        while i < len(s):
            if s[i:i+len(old)] == old:
                res += new
                i += len(old)
            else:
                res += s[i]
                i += 1
        
        return res

File: Data Structures & Algorithms/decode-ways/submission-20.java
class Solution {
    public int dfs(int i, String s, Map<Integer, Integer> memo) {
        // base case or already cached
        if (memo.containsKey(i)) {
            return memo.get(i);
        }
        // string starts with 0, its invalid
        if (s.charAt(i) == '0') {
            return 0;
        }

        // subproblem #1
        int res = dfs(i + 1, s, memo);

        // subproblem #2
        if (i + 1 < s.length()) {
            if (s.charAt(i) == '1' || s.charAt(i) == '2' && s.charAt(i + 1) < '7') {
                res += dfs(i + 2, s, memo);
            }
        }

        memo.put(i, res);
        return res;
    }

    public int numDecodings(String s) {
        Map<Integer, Integer> memo = new HashMap<>();
        memo.put(s.length(), 1);
        return dfs(0, s, memo);
    }
}
